#!/usr/bin/Rscript
#if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
#install.packages("stringr")
library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)
#' Simple SV type classifier
simpleEventType <- function(gr) {
  pgr = partner(gr)
  return(ifelse(seqnames(gr) != seqnames(pgr), "CTX", # inter-chromosomosal
    ifelse(strand(gr) == strand(pgr), "INV",
      ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "INS", # TODO: improve classification of complex events
        ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "DEL",
          "DUP")))))
}
# using the example in the GRIDSS /example directory

args <- commandArgs(trailingOnly = TRUE)
species <- args[1]
workingdir <- args[2]
# vcf <- readVcf("gridss.full.chr12.1527326.DEL1024.vcf", "hg19")
vcf <- readVcf(str_glue("{workingdir}/tmp.vcf"))
info(header(vcf)) = unique(as(rbind(as.data.frame(info(header(vcf))), data.frame(
	row.names=c("SIMPLE_TYPE"),
	Number=c("1"),
	Type=c("String"),
	Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))
gr <- breakpointRanges(vcf)
svtype <- simpleEventType(gr)
info(vcf)$SIMPLE_TYPE <- NA_character_
info(vcf[gr$sourceId])$SIMPLE_TYPE <- svtype
info(vcf[gr$sourceId])$SVLEN <- gr$svLen
info(vcf[gr$sourceId])$SVTYPE <- info(vcf[gr$sourceId])$SIMPLE_TYPE
# writeVcf(vcf, "chr12.1527326.DEL1024.sv.annotated.vcf") # generated by example/gridss.sh
writeVcf(vcf, str_glue("{workingdir}/{species}.gridss.tmp.annotated.vcf"))

# TODO: perform event filtering here
# By default, GRIDSS is very sensitive but this comes at the cost of a high false discovery rate
gr <- gr[gr$FILTER == "PASS" & partner(gr)$FILTER == "PASS"] # Remove low confidence calls

simplegr <- gr[simpleEventType(gr) %in% c("INS", "INV", "DEL", "DUP")]
simplebed <- data.frame(
	chrom=seqnames(simplegr),
	# call the centre of the homology/inexact interval
	start=as.integer((start(simplegr) + end(simplegr)) / 2),
	end=as.integer((start(partner(simplegr)) + end(partner(simplegr))) / 2),
	name=simpleEventType(simplegr),
	score=simplegr$QUAL,
	strand="."
)
# Just the lower of the two breakends so we don't output everything twice
simplebed <- simplebed[simplebed$start < simplebed$end,]
# write.table(simplebed, "chr12.1527326.DEL1024.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
write.table(simplebed, str_glue("{workingdir}/{species}.gridss.annotated.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE))

# args <- commandArgs(trailingOnly = TRUE)
# vcf <- readVcf("gridss.vcf", "hg19")
# gr <- breakpointRanges(vcf)
# bedpe <- data.frame(
#     chrom1=seqnames(gr),
#     start1=start(gr),
#     end1=end(gr),
#     chrom1=seqnames(partner(gr)),
#     start1=start(partner(gr)),
#     end1=end(partner(gr)),
#     name=names(gr),
#     score=gr$QUAL,
#     strand1=strand(gr),
#     strand2=strand(partner(gr))
#     )
# # Just the lower of the two breakends so we don't output everything twice
# bedpe <- bedpe[str_detect(bedpe$name, "gridss.+o"),]
# write.table(bedpe, "gridss.bedpe", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
